### PR TITLE
[herd] Lift memop for DC by VA instructions

### PR DESCRIPTION
**Pull request by relokin**
This change adds support for translation and other memory features for
DC instructions by VA. DC instructions by set and way bypass this.

Signed-off-by: Nikos Nikoleris <nikos.nikoleris@arm.com>

### CODE REVIEW

File: herd/AArch64Sem.ml [hunk 1]
```diff
@@ -1333,6 +1325,28 @@ module Make
...
+      let dc_loc op a ii =
+        let mk_act loc = Act.DC (op,Some loc) in
+        let loc = A.Location_global a in
+        M.mk_singleton_es (mk_act loc) ii
+
+      let do_dc op rd ii =
+        if AArch64Base.DC.sw op then
+          M.mk_singleton_es (Act.DC (op, None)) ii >>! B.Next
+        else begin
+            (* TODO: The size for DC should be a cache line *)
```
Comment (maranget) [inline herd/AArch64Sem.ml:1338]: Perhaps we can rely on the assumption: different variables are allocated to different cache lines? If you need t allocated variables to the same cache line, we can use arrays and then decide whether two cells are in the same cache line or not, assuming  alignment of arrays (_i.e._ alignment according to total array size).

Comment (relokin) [reply herd/AArch64Sem.ml:1338]: I think so. I am not very familiar with the specifics of mixed mode. 

In non-mixed mode: All accesses to a memory location have to be of a specific size, don't they (e.g, byte, half word, ...)? In the mixed mode then this assumption goes away. If that's the case, we can make the same assumption for DC instructions; in non-mixed mode, when we clean location `x` we use the size of all the previous or next memory instruction that accesses `x`.

In mixed mode: We have to be able to handle the case where `x` and `y` (or x[0] and x[1]) map into the same cache line, I think we have most of the mechanics in place, maybe we also need to be able to reason about allocation alignment.

Comment (maranget) [reply herd/AArch64Sem.ml:1338]: I do not know yet. Anyway I'll merge this PR.

File: herd/AArch64Sem.ml [hunk 2]
```diff
@@ -1333,6 +1325,28 @@ module Make
...
+            let mop _ac a = dc_loc op a ii in
+            lift_memop Dir.R
+              (fun ac ma _mv -> (* value fake here *)
+                if Access.is_physical ac then
+                  M.bind_ctrldata ma (mop ac)
```
Comment (maranget) [inline herd/AArch64Sem.ml:1343]: I wanted to merge this PR. But now I am not so sure about the control dependency. Can you please comment?

Comment (relokin) [reply herd/AArch64Sem.ml:1343]: This is very similar to what we do for regular load/store instructions https://github.com/herd/herdtools7/blob/cc5ae26620ff3191c215b10c932f695d3de1c7c8/herd/AArch64Sem.ml#L890. If I understand correctly, this has to do with the branching event we add between the `Rpte` event and the corresponding memory read/write event. DC instructions similarly to loads and stores may raise a fault if the PTE is invalid.

Comment (maranget) [reply herd/AArch64Sem.ml:1343]: I see that DC is assimilated to a read effect. This has some importance for the hardware update of the access and dirty bits. I'll explore this by some experiment before merging. Notice that documentation states that `DC CVAU` checks for read permission. Hence the `Dir.R` argument to `liftmem_op` looks appropriate.

Comment (relokin) [reply herd/AArch64Sem.ml:1343]: The clean or clean and invalidate DC variants (DC CVAU, DC CVAC, DC CIVAC) all require read permissions. There is one exception which is the invalidate DC IVAC instruction which requires write permissions. I need to check though if it sets the dirty bit.

Comment (maranget) [reply herd/AArch64Sem.ml:1343]: I have just checked on hardware that `DC CVAU` behaves as a load as regards hardware flag management...  I can check `DC IVAC`.

Comment (maranget) [reply herd/AArch64Sem.ml:1343]: As far as I have tested, `DC IVAC` require write permission. However if hardware management of the dirty bit is enabled, it no longer fails  but does *not* update the dirty bit...

Comment (relokin) [reply herd/AArch64Sem.ml:1346]: I think this version addresses the problem with DC IVAC, by raising a fault if it doesn't have write permissions but avoiding to update the DB if it's not set.

Comment (maranget) [reply herd/AArch64Sem.ml:1346]: You think right...

**Comment by maranget** [comment]
Merged, thanks.

**Comment by relokin** [comment]
Thanks Luc!

**Pull request merged**